public/app/plugins/datasource/mixed/MixedDatasource.ts

```diff
@@ -0,0 +1,124 @@
+import cloneDeep from 'lodash/cloneDeep';
+import groupBy from 'lodash/groupBy';
+import map from 'lodash/map';
+import flatten from 'lodash/flatten';
+
+import {
+  DataSourceApi,
+  DataQuery,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataStreamState,
+  DataStreamObserver,
+  DataSourceInstanceSettings,
+} from '@grafana/ui';
+import { LoadingState } from '@grafana/data';
+
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { getProcessedDataFrames } from 'app/features/dashboard/state/PanelQueryState';
+import { toDataQueryError } from 'app/features/dashboard/state/PanelQueryState';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+
+export const MIXED_DATASOURCE_NAME = '-- Mixed --';
+
+export class MixedDatasource extends DataSourceApi<DataQuery> {
+  constructor(instanceSettings: DataSourceInstanceSettings) {
+    super(instanceSettings);
+  }
+
+  async query(request: DataQueryRequest<DataQuery>, observer: DataStreamObserver): Promise<DataQueryResponse> {
+    // Remove any hidden or invalid queries
+    const queries = request.targets.filter(t => {
+      return !t.hide || t.datasource === MIXED_DATASOURCE_NAME;
+    });
+
+    if (!queries.length) {
+      return Promise.resolve({ data: [] }); // nothing
+    }
+
+    // When multiple datasources are used, merge their results
+    // after all queries have finished.  (<=6.3 behavior)
+    const sets = groupBy(queries, 'datasource');
```

Comment:
So this is one of the thing why I do not like this API very much. The behaviour of this function changes drastically based on whether there are multiple datasources or just one, which you as a caller do not see until you go into the params. This means you would need to always expect both promise and a stream just to be sure which then makes observer non optional for most cases which we cannot put in the type because then it would break backward compatibility. Right now this feels a bit like if the interface was `callFunction('functionName')` and you need to know internal implementation of that call to know what it will do and how to handle that.

Reply:
Would it make more sense if the api looked like this
```ts
async query(request: DataQueryRequest<DataQuery>): Observable<DataQueryResponse> 
```

This way you would always expect a stream of data, either from a single request or from multiple requests or live streaming.

Reply:
Or do we want to avoid the RxJs dependency?

Reply:
I think we tried something similar in https://github.com/grafana/grafana/pull/16714 

Using the old (v3-v6) streaming support was handled by returning a Subject. But think it's tricky to just return an observable of DataQueryResponse, think it would have to be more of a Observable<DataStreamState> or something inspired by that or expand DataQueryResponse  with all the things needed to track streaming events (key & source request).

Reply:
Well I think right now with the observer callback we already push the data there as an events with key for the source request if I am not mistaken. We could still be returning that in one observable, or possibly returning Observable of Observables, with one stream per source request that you can flatten if needed. Something like  `Observable<Observable<DataQueryResponse> | DataQueryResponse>` maybe, though you would need to assert the type on the way back based on what you sent in which I am not that fond of.

Reply:
Good points, I think that having more meta data like ((key & source request) and DataQueryResponse would be a simpler solution. The only thing I would be interested in as a subscriber of the stream are the actual "address" of the data not if there are one or multiple streams or if it's live streaming.